openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,352 @@
+        return error;
+    }
+
+    if (OCSP_CALLBACK_METHOD == NULL) {
+        tcn_ThrowException(env, "openssl_netty_tcnative_ocsp_callback: no callback method");
+        return error;
+    }
+
+    if (arg == NULL) {
+        tcn_ThrowException(env, "openssl_netty_tcnative_ocsp_callback: no callback args");
+        return error;
+    }
+
+    int response = 0;
+    jobject callback = (*env)->NewLocalRef(env, (jweak)arg);
+
+    if ((*env)->IsSameObject(env, callback, NULL) == JNI_FALSE) {
+        jlong ssl_ = P2J(ssl);
+        response = (int)(*env)->CallIntMethod(env, callback, OCSP_CALLBACK_METHOD, ssl_);
+
+        if ((*env)->ExceptionCheck(env)) {
+            // Use the given error code in the event of a JVM exception
+            response = error;
+        }
+    } else {
+        response = error;
+    }
+
+    (*env)->DeleteLocalRef(env, callback);
+    return response;
+}
+
+/**
+ * Client and Server callbacks use different error codes.
+ *
+ * @see openssl_netty_tcnative_ocsp_callback
+ */
+static int openssl_netty_tcnative_ocsp_client_callback(SSL *ssl, void *arg)
+{
+    return openssl_netty_tcnative_ocsp_callback(TCN_CLIENT_CALLBACK_ERROR, ssl, arg);
+}
+
+/**
+ * Client and Server callbacks use different error codes.
+ *
+ * @see openssl_netty_tcnative_ocsp_callback
+ */
+static int openssl_netty_tcnative_ocsp_server_callback(SSL *ssl, void *arg)
+{
+    return openssl_netty_tcnative_ocsp_callback(TCN_SERVER_CALLBACK_ERROR, ssl, arg);
+}
+#endif
+
+/**
+ * Returns true is OCSP stapling is available.
+ *
+ * NOTE: There can be basically three reasons why OCSP stapling
+ * may not be available:
+ *
+ * 1) It's outright not supported (e.g. OpenSSL < 1.0.0).
+ * 2) OpenSSL was explicitly built without OCSP stapling support.
+ * 3) It's supported but not implemented by us (e.g. BoringSSL)
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, isOcspAvailable)(TCN_STDARGS)
+{
+#if defined(TCN_NO_OCSP)
+    return JNI_FALSE;
+#elif defined(OPENSSL_IS_BORINGSSL)
+    // TODO: BoringSSL has OCSP support but uses a different API.
+    return JNI_FALSE;
+#else
+    return JNI_TRUE;
+#endif
+}
+
+/**
+ * Installs the OCSP stapling callback.
+ */
+TCN_IMPLEMENT_CALL(jobject, SSLContext, newOcspCallback)(TCN_STDARGS, jlong ctx, jboolean client, jobject callback)
+{
+#if defined(TCN_NO_OCSP)
+    tcn_ThrowException(e, "OCSP stapling is not supported");
+    return NULL;
+#elif defined(OPENSSL_IS_BORINGSSL)
+    // TODO: BoringSSL has OCSP support but uses a different API.
+    tcn_ThrowException(e, "OCSP stapling is not implemented");
+    return NULL;
+#else
+    long error = -1L;
+
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    if (c == NULL) {
+        tcn_ThrowException(e, "ctx is null");
+        return NULL;
+    }
+
+    //
+    // ATTENTION: We need to create a global reference of the callback.
+    //
+    jweak callbackRef = (*e)->NewWeakGlobalRef(e, callback);
```

Comment:
good idea ... @normanmaurer - we may want to consider doing something similar in other areas where we callback to java code.